src/client/array/dc_array.c

```diff
@@ -2373,12 +2316,16 @@ adjust_array_size_cb(tse_task_t *task, void *data)
 			if (dkey_val == 0)
 				continue;
 			/*
-			 * Punch the entire dkey since it's in a higher dkey
-			 * group than the intended size.
+			 * The dkey is higher than the adjustded size so we could punch it here.
+			 * But it's better to punch the extent so that the max_write for the object
+			 * doesn't get lost by aggregation.
 			 */
-			D_DEBUG(DB_IO, "Punching key: "DF_U64"\n", dkey_val);
-			rc = punch_key(args->oh, args->th, dkey_val,
-				       props->ptask);
+			D_DEBUG(DB_IO, "Punch full extent in key "DF_U64"\n",
+				dkey_val);
+			rc = punch_extent(args->oh, args->th,
+					  dkey_val, (daos_off_t)-1,
```

Comment:
Minor: It's not from this patch, but I think it could be easier for reader if you simply pass in the requested punch offset (which is 0 here) instead of '-1' here.  (the 'record_i + 1' in punch_extent() could be adjusted to use requested punch offset directly).

Reply:
Yeah, I didn't want to change more than the minimum in code I didn't write but if @mchaarawi agrees, I can take a stab at it.

Reply:
the idea behind this code is to punch everything above that record. it wasn't meant as a range punch.
then we fetch at that offset to determine record existence and then add that record if it doesn't exist.
i don't think it makes sense to change all this as part of this PR and would suggest to create a ticket and assign it to me if it bothers you that much ;-)

Reply:
It actually doesn't bother me, it bothered some developers (from SPDK team) working on the project of VOS based local fs. ;-)

Few days ago, on the regular meeting with them, they raised the question of why punch_extent() always start punch from "record_i + 1"? I've explained to them, so it's doesn't bother them anymore. :)

That's the reason I brought this up when I happen to see the change here. I don't think it's necessary make a change in current PR or open a new ticket.